Running tinygo test ./... fails

Running `tinygo test ./...` does not properly resolve all the packages recursively.

Seems to resolve recursively here on at least some packages.  Can you give a real-world example?  e.g.
```
$ git clone https://go.googlesource.com/image
$ cd image
$ tinygo test ./...
# golang.org/x/image/font/inconsolata
font/inconsolata/bold8x16.go:8:2: could not import golang.org/x/image/font/basicfont (Config.Importer.Import(golang.org/x/image/font/basicfont) returned nil but no error)
FAIL
```
(That's just an example of the kind of detail that would be great in this bug report)

In this very same repository, you can run

```
➜  tinygo git:(fixes_three_dots) ✗ tinygo test ./tests/testing/... 
error: Linking globals named 'main.tests': symbol multiply defined!
```
and after the change you get
```
➜  tinygo git:(fixes_three_dots) ✗ ~/Workspace/gource/bin/tinygo test ./tests/testing/... 
ld.lld: error: undefined symbol: _notARealFunction
>>> referenced by /var/folders/s7/qmqrxc894d55fymyt4x506300000gn/T/tinygo1047012886/main.o
error: failed to link /var/folders/s7/qmqrxc894d55fymyt4x506300000gn/T/tinygo1047012886/main: exit status 1
--- FAIL: TestFail (0.00s)
    fail
FAIL
FAIL    github.com/tinygo-org/tinygo/tests/testing/fail 0.326s
?       github.com/tinygo-org/tinygo/tests/testing/nothing      [no test files]
ok      github.com/tinygo-org/tinygo/tests/testing/pass 0.176s
FAIL
````

But all this started in the repository github.com/corazawaf/coraza where running `tinygo test ./...` only tested first package `github.com/corazawaf/coraza/v2`:

```
Run tinygo test ./...
# github.com/corazawaf/coraza/v2
Error: collection.go:23:2: could not import github.com/corazawaf/coraza/v2/types/variables (Config.Importer.Import(github.com/corazawaf/coraza/v2/types/variables) returned nil but no error)
Error: transaction.go:29:2: could not import github.com/corazawaf/coraza/v2/bodyprocessors (Config.Importer.Import(github.com/corazawaf/coraza/v2/bodyprocessors) returned nil but no error)
Error: transaction.go:30:2: could not import github.com/corazawaf/coraza/v2/loggers (Config.Importer.Import(github.com/corazawaf/coraza/v2/loggers) returned nil but no error)
Error: transaction.go:3[4](https://github.com/corazawaf/coraza/runs/6661735281?check_suite_focus=true#step:6:5):7: could not import github.com/corazawaf/coraza/v2/utils/url (Config.Importer.Import(github.com/corazawaf/coraza/v2/utils/url) returned nil but no error)
FAIL
Error: transaction.go:34[5](https://github.com/corazawaf/coraza/runs/6661735281?check_suite_focus=true#step:6:6):25: invalid operation: cannot compare tx.bodyProcessor != nil (operator != not defined on untyped nil)
Error: Process completed with exit code 1.
```

--- a/Makefile
+++ b/Makefile
@@ -377,6 +377,8 @@ tinygo-baremetal:
 .PHONY: smoketest
 smoketest:
 	$(TINYGO) version
+	# regression test for #2892
+	cd tests/testing/recurse && ($(TINYGO) test ./... > recurse.log && cat recurse.log && test $$(wc -l < recurse.log) = 2 && rm recurse.log)
 	# compile-only platform-independent examples
 	cd tests/text/template/smoke && $(TINYGO) test -c && rm -f smoke.test
 	# regression test for #2563

--- a/builder/build.go
+++ b/builder/build.go
@@ -193,7 +193,7 @@ func Build(pkgName, outpath string, config *compileopts.Config, action func(Buil
 	defer machine.Dispose()
 
 	// Load entire program AST into memory.
-	lprogram, err := loader.Load(config, []string{pkgName}, config.ClangHeaders, types.Config{
+	lprogram, err := loader.Load(config, pkgName, config.ClangHeaders, types.Config{
 		Sizes: compiler.Sizes(machine),
 	})
 	if err != nil {

--- a/compiler/compiler_test.go
+++ b/compiler/compiler_test.go
@@ -109,7 +109,7 @@ func TestCompiler(t *testing.T) {
 			defer machine.Dispose()
 
 			// Load entire program AST into memory.
-			lprogram, err := loader.Load(config, []string{"./testdata/" + tc.file}, config.ClangHeaders, types.Config{
+			lprogram, err := loader.Load(config, "./testdata/"+tc.file, config.ClangHeaders, types.Config{
 				Sizes: Sizes(machine),
 			})
 			if err != nil {

--- a/loader/loader.go
+++ b/loader/loader.go
@@ -104,7 +104,7 @@ type EmbedFile struct {
 // Load loads the given package with all dependencies (including the runtime
 // package). Call .Parse() afterwards to parse all Go files (including CGo
 // processing, if necessary).
-func Load(config *compileopts.Config, inputPkgs []string, clangHeaders string, typeChecker types.Config) (*Program, error) {
+func Load(config *compileopts.Config, inputPkg string, clangHeaders string, typeChecker types.Config) (*Program, error) {
 	goroot, err := GetCachedGoroot(config)
 	if err != nil {
 		return nil, err
@@ -133,7 +133,7 @@ func Load(config *compileopts.Config, inputPkgs []string, clangHeaders string, t
 	if config.TestConfig.CompileTestBinary {
 		extraArgs = append(extraArgs, "-test")
 	}
-	cmd, err := List(config, extraArgs, inputPkgs)
+	cmd, err := List(config, extraArgs, []string{inputPkg})
 	if err != nil {
 		return nil, err
 	}

--- a/main.go
+++ b/main.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"bufio"
 	"bytes"
 	"context"
 	"encoding/json"
@@ -1252,6 +1253,35 @@ func parseGoLinkFlag(flagsString string) (map[string]map[string]string, error) {
 	return map[string]map[string]string(globalVarValues), nil
 }
 
+// getListOfPackages returns a standard list of packages for a given list that might
+// include wildards using `go list`.
+// For example [./...] => ["pkg1", "pkg1/pkg12", "pkg2"]
+func getListOfPackages(pkgs []string, options *compileopts.Options) ([]string, error) {
+	config, err := builder.NewConfig(options)
+	if err != nil {
+		return nil, err
+	}
+	cmd, err := loader.List(config, nil, pkgs)
+	if err != nil {
+		return nil, fmt.Errorf("failed to run `go list`: %w", err)
+	}
+	outputBuf := bytes.NewBuffer(nil)
+	cmd.Stdout = outputBuf
+	cmd.Stderr = os.Stderr
+	err = cmd.Run()
+	if err != nil {
+		return nil, err
+	}
+
+	var pkgNames []string
+	sc := bufio.NewScanner(outputBuf)
+	for sc.Scan() {
+		pkgNames = append(pkgNames, sc.Text())
+	}
+
+	return pkgNames, nil
+}
+
 func main() {
 	if len(os.Args) < 2 {
 		fmt.Fprintln(os.Stderr, "No command-line arguments supplied.")
@@ -1487,14 +1517,21 @@ func main() {
 		if len(pkgNames) == 0 {
 			pkgNames = []string{"."}
 		}
-		if outpath != "" && len(pkgNames) > 1 {
+
+		explicitPkgNames, err := getListOfPackages(pkgNames, options)
+		if err != nil {
+			fmt.Printf("cannot resolve packages: %v\n", err)
+			os.Exit(1)
+		}
+
+		if outpath != "" && len(explicitPkgNames) > 1 {
 			fmt.Println("cannot use -o flag with multiple packages")
 			os.Exit(1)
 		}
 
 		fail := make(chan struct{}, 1)
 		var wg sync.WaitGroup
-		bufs := make([]testOutputBuf, len(pkgNames))
+		bufs := make([]testOutputBuf, len(explicitPkgNames))
 		for i := range bufs {
 			bufs[i].done = make(chan struct{})
 		}
@@ -1520,7 +1557,7 @@ func main() {
 		// Build and run the tests concurrently.
 		// This uses an additional semaphore to reduce the memory usage.
 		testSema := make(chan struct{}, cap(options.Semaphore))
-		for i, pkgName := range pkgNames {
+		for i, pkgName := range explicitPkgNames {
 			pkgName := pkgName
 			buf := &bufs[i]
 			testSema <- struct{}{}

--- a/main_test.go
+++ b/main_test.go
@@ -13,6 +13,7 @@ import (
 	"io/ioutil"
 	"os"
 	"os/exec"
+	"reflect"
 	"regexp"
 	"runtime"
 	"strings"
@@ -518,6 +519,46 @@ func ioLogger(t *testing.T, wg *sync.WaitGroup) io.WriteCloser {
 	}()
 
 	return w
+}
+
+func TestGetListOfPackages(t *testing.T) {
+	opts := optionsFromTarget("", sema)
+	tests := []struct {
+		pkgs          []string
+		expectedPkgs  []string
+		expectesError bool
+	}{
+		{
+			pkgs: []string{"./tests/testing/recurse/..."},
+			expectedPkgs: []string{
+				"github.com/tinygo-org/tinygo/tests/testing/recurse",
+				"github.com/tinygo-org/tinygo/tests/testing/recurse/subdir",
+			},
+		},
+		{
+			pkgs: []string{"./tests/testing/pass"},
+			expectedPkgs: []string{
+				"github.com/tinygo-org/tinygo/tests/testing/pass",
+			},
+		},
+		{
+			pkgs:          []string{"./tests/testing"},
+			expectesError: true,
+		},
+	}
+
+	for _, test := range tests {
+		actualPkgs, err := getListOfPackages(test.pkgs, &opts)
+		if err != nil && !test.expectesError {
+			t.Errorf("unexpected error: %v", err)
+		} else if err == nil && test.expectesError {
+			t.Error("expected error, but got none")
+		}
+
+		if !reflect.DeepEqual(test.expectedPkgs, actualPkgs) {
+			t.Errorf("expected two slices to be equal, expected %v got %v", test.expectedPkgs, actualPkgs)
+		}
+	}
 }
 
 // This TestMain is necessary because TinyGo may also be invoked to run certain

--- a/tests/testing/recurse/subdir/subdir_test.go
+++ b/tests/testing/recurse/subdir/subdir_test.go
@@ -1,0 +1,6 @@
+package subdir
+
+import "testing"
+
+func TestSubdir(t *testing.T) {
+}

--- a/tests/testing/recurse/top_test.go
+++ b/tests/testing/recurse/top_test.go
@@ -1,0 +1,6 @@
+package top
+
+import "testing"
+
+func TestTop(t *testing.T) {
+}

--- a/transform/transform_test.go
+++ b/transform/transform_test.go
@@ -146,7 +146,7 @@ func compileGoFileForTesting(t *testing.T, filename string) llvm.Module {
 	defer machine.Dispose()
 
 	// Load entire program AST into memory.
-	lprogram, err := loader.Load(config, []string{filename}, config.ClangHeaders, types.Config{
+	lprogram, err := loader.Load(config, filename, config.ClangHeaders, types.Config{
 		Sizes: compiler.Sizes(machine),
 	})
 	if err != nil {

